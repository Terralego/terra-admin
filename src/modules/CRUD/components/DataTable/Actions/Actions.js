import React, { useMemo, useContext, useCallback } from 'react';
import PropTypes from 'prop-types';

import { useTranslation } from 'react-i18next';

import {
  Button,
  Intent,
  Position,
  Popover,
  PopoverInteractionKind,
} from '@blueprintjs/core';
import ColumnsSelector from '@terralego/core/modules/Table/components/ColumnsSelector';
import { UserSettingsContext, TABLE_MINIFIED, TABLE_MEDIUM, TABLE_FULL } from '../../../services/UserSettingsProvider';

import Search from '../Search';

const resizeButtonsProps = [
  { action: TABLE_FULL, icon: 'maximize' },
  { action: TABLE_MEDIUM, icon: 'minimize' },
  { action: TABLE_MINIFIED, icon: 'minus', text: 'hide' },
];

const Actions = ({
  columns,
  onChange,
  onHeaderChange,
}) => {
  const { t } = useTranslation();
  const {
    setTableSize,
    tableSize = TABLE_MEDIUM,
  } = useContext(UserSettingsContext);

  const popoverProps = useMemo(() => ({
    interactionKind: PopoverInteractionKind.HOVER,
    position: tableSize === TABLE_FULL ? Position.BOTTOM : Position.TOP,
    boundary: 'window',
  }), [tableSize]);

  const displayTableFilters = useMemo(() => (
    Boolean(columns.length) && tableSize !== TABLE_MINIFIED
  ), [columns.length, tableSize]);

  const handleChange = useCallback(props => {
    onChange(props);
    onHeaderChange(props);
  }, [onChange, onHeaderChange]);

  return (
    <div className="table-header__actions">
      {displayTableFilters && (
        <>
          <Search />
          <Popover
            content={t('CRUD.table.filterProps')}
            {...popoverProps}
          >
            <ColumnsSelector
              columns={columns}
              onChange={handleChange}
              position={Position.LEFT}
              locales={{
                displayAllColumns: t('CRUD.table.columnsDisplay'),
                hideAllColumns: t('CRUD.table.columnsHide'),
              }}
            />
          </Popover>
        </>
      )}
      <Popover
        content={(
          <div className="table-header__resize">
            {resizeButtonsProps.map(({ action, icon, text = icon }) => (
              <Button
                active={tableSize === action}
                key={action}
                icon={icon}
                minimal
                onClick={() => setTableSize(action)}
                text={t(`CRUD.table.${text}`)}
              />
            ))}
          </div>
        )}
        {...popoverProps}
      >
        <Button
          icon="arrows-vertical"
          minimal
          intent={Intent.PRIMARY}
        />
      </Popover>
    </div>
  );
};

Actions.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.object),
  displayAddFeature: PropTypes.bool,
  layerName: PropTypes.string,
  onChange: PropTypes.func,
  onHeaderChange: PropTypes.func,
};

Actions.defaultProps = {
  columns: [],
  displayAddFeature: false,
  layerName: '',
  onChange: () => {},
  onHeaderChange: () => {},
};

export default Actions;
