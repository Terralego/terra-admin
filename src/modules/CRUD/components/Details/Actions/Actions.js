import React from 'react';
import { NavLink } from 'react-router-dom';
import { Icon, Button, Popover, H5, PopoverInteractionKind, Classes } from '@blueprintjs/core';

import { generateURI } from '../../../config';
import { toast } from '../../../../../utils/toast';

class Actions extends React.Component {
  deleteFeature = async () => {
    const {
      paramLayer,
      paramId,
      layer: { id: layerId },
      deleteFeature,
      history: { push },
      t,
    } = this.props;

    const deleted = await deleteFeature(layerId, paramId);

    toast.displayToaster(
      deleted ? { id: deleted } : { },
      t('CRUD.details.successDeleteFeature'),
      t('CRUD.details.failDeleteFeature'),
    );

    if (!deleted) {
      return;
    }

    push(generateURI('layer', { layer: paramLayer }));
  }

  render () {
    const {
      paramLayer,
      paramId,
      displayUpdate,
      displayDelete,
      displayCancel,
      children,
      t,
    } = this.props;
    return (
      <div className="details__actions">
        {displayDelete && (
          <Popover
            popoverClassName={Classes.POPOVER_CONTENT_SIZING}
            interactionKind={PopoverInteractionKind.CLICK}
            className="details__actions-delete"
            content={(
              <div className="details__confirm">
                <H5>{t('CRUD.details.confirmDeletion')}</H5>
                <p>{t('CRUD.details.confirmDeletionText')}</p>
                <div className="details__confirm-content">
                  <Button
                    className={Classes.POPOVER_DISMISS}
                    text={t('CRUD.details.cancel')}
                  />
                  <Button
                    intent="danger"
                    className={Classes.POPOVER_DISMISS}
                    text={t('CRUD.details.delete')}
                    onClick={this.deleteFeature}
                  />
                </div>
              </div>
            )}
          >
            <Button icon="trash" intent="danger" text={t('CRUD.details.delete')} />
          </Popover>
        )}
        {displayUpdate && (
          <NavLink to={generateURI('layer', { layer: paramLayer, id: paramId, action: 'update' })}>
            <span className="bp3-button bp3-intent-primary">
              <Icon icon="edit" />
              <span className="bp3-button-text">{t('CRUD.details.update')}</span>
            </span>
          </NavLink>
        )}
        {displayCancel && (
          <NavLink to={generateURI('layer', { layer: paramLayer, id: paramId })}>
            <span className="bp3-button">
              <Icon icon="undo" />
              <span className="bp3-button-text">{t('CRUD.details.cancel')}</span>
            </span>
          </NavLink>
        )}
        {!!children && (
          <>{children}</>
        )}
      </div>
    );
  }
}

export default Actions;
