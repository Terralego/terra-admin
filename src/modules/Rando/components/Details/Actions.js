import React from 'react';
import { NavLink } from 'react-router-dom';
import { Icon, Button, Popover, H5, PopoverInteractionKind, Classes } from '@blueprintjs/core';
import { withNamespaces } from 'react-i18next';
import { withRouter } from 'react-router';

import { connectRandoProvider } from '../../services/RandoProvider';
import { generateURI } from '../../config';
import { toast } from '../../../../utils/toast';

class Actions extends React.Component {
  deleteFeature = () => {
    const { layer, id, deleteFeature, history: { push }, t } = this.props;
    deleteFeature(layer, id);
    toast.displayToaster(
      { id },
      t('rando.details.successDeleteFeature'),
      t('rando.details.failDeleteFeature'),
    );
    push(generateURI('layer', { layer }));
  }

  render () {
    const { layer, id, displayUpdate, displayDelete, displayCancel, t } = this.props;
    return (
      <div className="details__actions">
        {displayUpdate && (
        <NavLink to={generateURI('layer', { layer, id, action: 'update' })}>
          <span className="bp3-button">
            <Icon icon="edit" />
            <span className="bp3-button-text">{t('rando.details.update')}</span>
          </span>
        </NavLink>
        )}
        {displayCancel && (
          <NavLink to={generateURI('layer', { layer, id })}>
            <span className="bp3-button">
              <Icon icon="undo" />
              <span className="bp3-button-text">{t('rando.details.cancel')}</span>
            </span>
          </NavLink>
        )}
        {displayDelete && (
          <Popover
            popoverClassName={Classes.POPOVER_CONTENT_SIZING}
            interactionKind={PopoverInteractionKind.CLICK}
            content={(
              <div className="details__confirm">
                <H5>{t('rando.details.confirmDeletion')}</H5>
                <p>{t('rando.details.confirmDeletionText')}</p>
                <div className="details__confirm-content">
                  <Button
                    className={Classes.POPOVER_DISMISS}
                    text={t('rando.details.cancel')}
                  />
                  <Button
                    intent="danger"
                    className={Classes.POPOVER_DISMISS}
                    text={t('rando.details.delete')}
                    onClick={this.deleteFeature}
                  />
                </div>
              </div>
            )}
          >
            <Button icon="trash" intent="danger" text={t('rando.details.delete')} />
          </Popover>
        )}
      </div>
    );
  }
}

export default withRouter(connectRandoProvider(
  'deleteFeature',
)(withNamespaces()(Actions)));
