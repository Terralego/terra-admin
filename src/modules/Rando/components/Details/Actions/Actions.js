import React from 'react';
import { NavLink } from 'react-router-dom';
import { Icon, Button, Popover, H5, PopoverInteractionKind, Classes } from '@blueprintjs/core';

import { generateURI } from '../../../config';
import { toast } from '../../../../../utils/toast';

class Actions extends React.Component {
  deleteFeature = () => {
    const {
      paramLayer,
      paramId,
      layer: { id: layerId },
      deleteFeature,
      history: { push },
      t,
    } = this.props;

    deleteFeature(layerId, paramId);
    toast.displayToaster(
      { id: paramId },
      t('rando.details.successDeleteFeature'),
      t('rando.details.failDeleteFeature'),
    );
    push(generateURI('layer', { layer: paramLayer }));
  }

  render () {
    const {
      paramLayer,
      paramId,
      displayUpdate,
      displayDelete,
      displayCancel,
      children,
      t,
    } = this.props;
    return (
      <div className="details__actions">
        {displayDelete && (
          <Popover
            popoverClassName={Classes.POPOVER_CONTENT_SIZING}
            interactionKind={PopoverInteractionKind.CLICK}
            className="details__actions-delete"
            content={(
              <div className="details__confirm">
                <H5>{t('rando.details.confirmDeletion')}</H5>
                <p>{t('rando.details.confirmDeletionText')}</p>
                <div className="details__confirm-content">
                  <Button
                    className={Classes.POPOVER_DISMISS}
                    text={t('rando.details.cancel')}
                  />
                  <Button
                    intent="danger"
                    className={Classes.POPOVER_DISMISS}
                    text={t('rando.details.delete')}
                    onClick={this.deleteFeature}
                  />
                </div>
              </div>
            )}
          >
            <Button icon="trash" intent="danger" text={t('rando.details.delete')} />
          </Popover>
        )}
        {displayUpdate && (
          <NavLink to={generateURI('layer', { layer: paramLayer, id: paramId, action: 'update' })}>
            <span className="bp3-button bp3-intent-primary">
              <Icon icon="edit" />
              <span className="bp3-button-text">{t('rando.details.update')}</span>
            </span>
          </NavLink>
        )}
        {displayCancel && (
          <NavLink to={generateURI('layer', { layer: paramLayer, id: paramId })}>
            <span className="bp3-button">
              <Icon icon="undo" />
              <span className="bp3-button-text">{t('rando.details.cancel')}</span>
            </span>
          </NavLink>
        )}
        {!!children && (
          <>{children}</>
        )}
      </div>
    );
  }
}

export default Actions;
