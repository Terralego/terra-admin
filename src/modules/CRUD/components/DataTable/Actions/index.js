import { withTranslation } from 'react-i18next';
import { withTableSize } from '../../../services/UserSettingsProvider';
import compose from '../../../../../utils/compose';

import Actions from './Actions';

export default compose(
  withTranslation(),
  withTableSize(),
)(Actions);
