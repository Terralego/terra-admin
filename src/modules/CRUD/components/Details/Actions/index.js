import { withTranslation } from 'react-i18next';
import { withRouter } from 'react-router-dom';
import { connectAuthProvider } from '@terralego/core/modules/Auth';

import { connectAppProvider } from '../../../../../components/AppProvider';
import { connectCRUDProvider } from '../../../services/CRUDProvider';
import { getView } from '../../../services/CRUD';
import compose from '../../../../../utils/compose';

import Actions from './Actions';

const appProvider = ({
  env: { modules: { CRUD: { settings } } },
}) => ({
  settingsEndpoint: settings,
});

const authProvider = ({
  authenticated,
  user,
}, { displayDelete }) => {
  const permissions = authenticated ? user.permissions : [];
  return {
    displayDelete: displayDelete && permissions.includes('can_delete_feature'),
  };
};

const CRUDPRovider = ({
  getSettings,
  settings,
  deleteFeature,
}, {
  match: { params: { layer } },
}) => ({
  getSettings,
  deleteFeature,
  view: getView(settings, layer),
});

export default compose(
  withRouter,
  connectAppProvider(appProvider),
  connectAuthProvider(authProvider),
  connectCRUDProvider(CRUDPRovider),
  withTranslation(),
)(Actions);
