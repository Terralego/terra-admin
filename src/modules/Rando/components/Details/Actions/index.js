import { withNamespaces } from 'react-i18next';
import { withRouter } from 'react-router';
import { connectAuthProvider } from '@terralego/core/modules/Auth';

import { connectRandoProvider } from '../../../services/RandoProvider';
import Actions from './Actions';

export default connectAuthProvider(({
  user,
}, { displayUpdate, displayDelete }) => {
  const { permissions = [] } = user;
  return {
    displayDelete: displayDelete && permissions.includes('terra.delete_feature'),
    displayUpdate: displayUpdate && permissions.includes('terra.change_feature'),
  };
})(
  withRouter(
    connectRandoProvider(({
      layersList,
      deleteFeature,
    }, {
      match: { params: { layer } },
    }) => ({
      deleteFeature,
      layer: layersList.find(({ name }) => name === layer),
    }))(withNamespaces()(Actions)),
  ),
);
