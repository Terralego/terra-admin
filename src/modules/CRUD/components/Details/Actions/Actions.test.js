import React from 'react';
import renderer from 'react-test-renderer';

import Actions from './Actions';

jest.mock('react-router-dom', () => ({
  NavLink: ({ children }) => <div>{children}</div>,
}));

jest.mock('@blueprintjs/core', () => ({
  Icon: ({ icon }) => <li>{icon}</li>,
  Button: ({ children }) => <div>{children}</div>,
  Popover: ({ children }) => <div>{children}</div>,
  H5: ({ children }) => <h5>{children}</h5>,
  PopoverInteractionKind: {
    CLICK: 'click',
  },
  Classes: {
    POPOVER_CONTENT_SIZING: 'POPOVER_CONTENT_SIZING',
  },
}));

jest.mock('../../../config', () => ({
  generateURI: jest.fn(),
}));

jest.mock('../../../../../utils/toast', () => ({
  toast: {
    displayToaster: bool => <div>{bool ? 'Success' : 'Fail'}</div>,
  },
}));

let props;
beforeEach(() => {
  props = {
    paramLayer: 'LayerFoo',
    paramId: 'featureFoo',
    displayUpdate: true,
    displayDelete: true,
    displayCancel: true,
    children: <div>Children component</div>,
    t: key => key,
    layer: {
      id: 10,
    },
    history: {
      push: jest.fn(),
    },
    deleteFeature: () => 'foo',
    getSettings: jest.fn(),
  };
});


it('should render correctly', () => {
  const tree = renderer.create((
    <Actions
      {...props}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should delete feature', async () => {
  const instance = new Actions({ ...props });
  instance.deleteFeature();
  await true;
  expect(instance.props.getSettings).toHaveBeenCalled();
});

it('should not crash when no deleting feature', async () => {
  const instance = new Actions({ ...props, deleteFeature: () => undefined });
  instance.deleteFeature();
  await true;
  expect(instance.props.getSettings).not.toHaveBeenCalled();
});
