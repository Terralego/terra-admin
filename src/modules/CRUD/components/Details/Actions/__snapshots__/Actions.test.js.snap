// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="details__actions"
>
  <div>
    <div />
  </div>
  <div>
    Children component
  </div>
</div>
`;
